### PR TITLE
Add API to to check if a user has a specific scope (#492)

### LINKED ISSUE #492: Add API to to check if a user has a specific scope
in a specific space/resource.

### PR DESCRIPTION
**Pull request by xcoulon**
Added endpoint as /api/resources/:resourceID/scopes/:scopeName

Also:
- refactored `controller/resource_roles_blackbox_test.go` by
  grouping in subtests

Fixes #492

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-599-1`

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-599-2`

File: authorization/role/service/role_management_service.go [hunk 1]
```diff
@@ -46,6 +46,11 @@ func (s *roleManagementServiceImpl) ListByResource(ctx context.Context, currentI
 	return s.Repositories().IdentityRoleRepository().FindIdentityRolesByResource(ctx, resourceID, false)
 }
 
+// ListByResource lists all identity roles for the resource if the current user has permissions to view the roles
+func (s *roleManagementServiceImpl) ListByIdentityAndResource(ctx context.Context, currentIdentity uuid.UUID, resourceID string) ([]role.RoleDescriptor, error) {
+	return s.ListByIdentityAndResource(ctx, currentIdentity, resourceID)
```
Comment (alexeykazakov) [inline authorization/role/service/role_management_service.go]: This method is
- broken (looks like an infinite recursive loop?)
- not covered by tests
- not used anywhere :)

Wonder if it was some temporal experiments you forgot to clean up.

**Comment by alexeykazakov** [comment]
@sbose78 @xcoulon @sbryzak I have concerns regarding exposing HasScope() method to a public endpoint.
Don't want clients to relay on that endpoint instead of using RPT tokens because this method doesn't support any cache/optimization. It was designed to be used internally. So Auth won't be able to control usage of that heavy and slow method. But I also understand that implementing the PRT flow (so it's easy to be used by clients) will take awhile.

**Comment by xcoulon** [comment]
@alexeykazakov good point for the lack of cache/optimization here. Maybe there's something to add in the `PermissionService` layer to benefit from caching, too?

**Comment by alexeykazakov** [comment]
@sbryzak is working on a cache for RPT tokens. Wondering if we could reuse it here as well if you guys believe we really do need to expose that endpoint :)

**Comment by sbose78** [comment]
Exactly. That was on my mind as well :)

Here's how we could go about it -

1. Expose the API as here. ( Easy to do, and makes our authz service usable, but comes with a cost)

2. Finish the RPT workflow ( Complex to do, but optimizes our service - and with any such optimization things might take a while to mature )

3. Utilize the optimization in the exposed Rest API as well.

That way our authorization service will give the resource servers the option to either use the RPT token workflow or the API endpoint.

**Comment by alexeykazakov** [comment]
@sbose78 good plan. +1

**Comment by xcoulon** [comment]
@sbose78 ok, let's do that way. I was also thinking about the RPT workflow and the DB caching, but I'm not sure if we'll be able to re-use it as-is or if it'll need some refactoring. 

Also, in its current form, the PR provides a new endpoint to check if the user has a specific scope in a requested resource, i.e., something like `/api/resources/:resourceID/scopes/:scopeName` and it returns a JSON-API response with a boolean attribute (`hasScope`). This is the way we can directly expose the `HasScope` service method. If we want to list all scopes for the current user in the request resource as was suggested in https://github.com/fabric8-services/fabric8-auth/issues/492#issuecomment-389024147, then we'll need another SQL request which @sbryzak is already working on as part of the RPT workflow.

**Comment by sbose78** [comment]
@xcoulon I think this API is nice. Just checking if a user has the needed scope is enough. Let's start with this.

**Comment by xcoulon** [comment]
thanks for your feedback, @sbose78. Could you approve the PR then, please ?

File: authorization/role/service/role_management_service.go (hunk 1 above)
Comment (sbose78) [reply authorization/role/service/role_management_service.go]: @xcoulon , have you taken care of this?

Comment (xcoulon) [reply authorization/role/service/role_management_service.go]: oh.... my bad! Let me take care of it now!

**Comment by fabric8cd** [comment]
@xcoulon snapshot fabric8-auth image is available for testing.  `docker pull docker.io/fabric8/fabric8-auth:SNAPSHOT-PR-599-3`

**Comment by xcoulon** [comment]
[test]

File: authorization/role/service/role_management_service.go (hunk 1 above)
Comment (xcoulon) [reply authorization/role/service/role_management_service.go]: @sbose78 it's now been taken care of. See https://github.com/fabric8-services/fabric8-auth/pull/599/commits/3f93465f00366a503a8542d1fe357310213be32c

**Comment by xcoulon** [comment]
[test]

**Comment by xcoulon** [comment]
[test]

**Comment by xcoulon** [comment]
thanks, @sbose78 !

File: design/role.go [hunk 2]
```diff
@@ -87,6 +87,21 @@ var _ = a.Resource("resource_roles", func() {
 		a.Response(d.Unauthorized, JSONAPIErrors)
 		a.Response(d.Conflict, JSONAPIErrors)
 	})
+	a.Action("hasScope", func() {
```
Comment (sbose78) [inline design/role.go]: You've missed the jwt protection here, @xcoulon .

Comment (xcoulon) [reply design/role.go]: ouch! Let me fix that immediately!

**Pull request merged**